File: sdk-python/conduit/policy.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from .client import Conduit, default_client

if TYPE_CHECKING:
    from .agent import Agent


@dataclass
class _PolicyState:
    max_per_transaction: int | None = None
    max_per_hour: int | None = None
    max_per_day: int | None = None
    max_per_minute_count: int = 60
    allowlist: list[str] | None = None
    blocklist: list[str] | None = None
    require_memo: bool = False
    enabled: bool = True


class Policy:
    """Spending policy bound to an Agent.

    Accessed via `agent.policy.attach(...)` matching the website snippet:

        agent.policy.attach(
            max_per_hour=10_000,
            allowlist=["02beef...", "02dead..."],
        )
    """

    def __init__(self, agent: "Agent", client: Conduit | None = None) -> None:
        self._agent = agent
        self._client = client or default_client()
        self._state = _PolicyState()

    def attach(
        self,
        *,
        max_per_transaction: int | None = None,
        max_per_hour: int | None = None,
        max_per_day: int | None = None,
        max_per_minute_count: int = 60,
        allowlist: list[str] | None = None,
        blocklist: list[str] | None = None,
        require_memo: bool = False,
        enabled: bool = True,
    ) -> "Policy":
        payload: dict[str, Any] = {
            "max_per_transaction": max_per_transaction,
            "max_per_hour": max_per_hour,
            "max_per_day": max_per_day,
            "max_per_minute_count": max_per_minute_count,
            "allowlist": list(allowlist) if allowlist else None,
            "blocklist": list(blocklist) if blocklist else None,
            "require_memo": require_memo,
            "enabled": enabled,
        }
        data = self._client.post(f"/v1/agents/{self._agent.id}/policy", json=payload)
        self._hydrate(data)
        return self

    def fetch(self) -> "Policy":
        data = self._client.get(f"/v1/agents/{self._agent.id}/policy")
        self._hydrate(data)
        return self

    def update(self, **kwargs: Any) -> "Policy":
        return self.attach(**kwargs)

    def remove(self) -> None:
        self._client.delete(f"/v1/agents/{self._agent.id}/policy")
        self._state = _PolicyState()

    def _hydrate(self, data: dict[str, Any]) -> None:
        self._state = _PolicyState(
            max_per_transaction=data.get("max_per_transaction"),
            max_per_hour=data.get("max_per_hour"),
            max_per_day=data.get("max_per_day"),
            max_per_minute_count=data.get("max_per_minute_count", 60),
            allowlist=data.get("allowlist") or None,
            blocklist=data.get("blocklist") or None,
            require_memo=bool(data.get("require_memo")),
            enabled=bool(data.get("enabled", True)),
        )

    def __getattr__(self, name: str) -> Any:
        if name.startswith("_"):
            raise AttributeError(name)
        return getattr(self._state, name)

File: sdk-js/src/policy.ts
import { Conduit, defaultClient } from "./client.js";
import type { Agent } from "./agent.js";
import type { PolicyAttachOptions } from "./types.js";

export class Policy {
  maxPerTransaction: number | null = null;
  maxPerHour: number | null = null;
  maxPerDay: number | null = null;
  maxPerMinuteCount = 60;
  allowlist: string[] = [];
  blocklist: string[] = [];
  requireMemo = false;
  enabled = true;

  constructor(
    private readonly agent: Agent,
    private readonly client: Conduit = defaultClient(),
  ) {}

  async attach(opts: PolicyAttachOptions = {}): Promise<Policy> {
    const data = await this.client.post<Record<string, unknown>>(
      `/v1/agents/${this.agent.id}/policy`,
      {
        max_per_transaction: opts.maxPerTransaction ?? null,
        max_per_hour: opts.maxPerHour ?? null,
        max_per_day: opts.maxPerDay ?? null,
        max_per_minute_count: opts.maxPerMinuteCount ?? 60,
        allowlist: opts.allowlist ?? null,
        blocklist: opts.blocklist ?? null,
        require_memo: opts.requireMemo ?? false,
        enabled: opts.enabled ?? true,
      },
    );
    this.hydrate(data);
    return this;
  }

  async fetch(): Promise<Policy> {
    const data = await this.client.get<Record<string, unknown>>(
      `/v1/agents/${this.agent.id}/policy`,
    );
    this.hydrate(data);
    return this;
  }

  async remove(): Promise<void> {
    await this.client.delete(`/v1/agents/${this.agent.id}/policy`);
  }

  private hydrate(d: Record<string, unknown>): void {
    this.maxPerTransaction = (d.max_per_transaction as number | null) ?? null;
    this.maxPerHour = (d.max_per_hour as number | null) ?? null;
    this.maxPerDay = (d.max_per_day as number | null) ?? null;
    this.maxPerMinuteCount = (d.max_per_minute_count as number) ?? 60;
    this.allowlist = (d.allowlist as string[]) ?? [];
    this.blocklist = (d.blocklist as string[]) ?? [];
    this.requireMemo = Boolean(d.require_memo);
    this.enabled = d.enabled !== false;
  }
}
